README.md
UniMazeGenEX DingoMode 2015
==========================

UniMazeGenEX DingoMode 2015 is a toy built on pygame and pygcurse that generates random displays of ASCII drawing characters. It features various controls to try to give the user a lot of power, such as changing the colours, the range of characters being drawn, and taking screenshots.

Usage
=====
Requires python 2, pygame, and pygcurse (please get pygcurse from here: https://github.com/asweigart/pygcurse).  
python runme.py

Controls
========
`1234567890 toggles various characters for display  
~!@#$%^&*() selects only one character to display, corresponding to the above keys  
SPACE toggles space character  
TAB toggles 4 * space character  
ENTER toggles 8 * space character  
BACKSPACE clears all characters to be printed  
B changes current character set (currently Double Pipes, Single Pipes, Card Suits, Numbers, and Shift Row)

-+ decreases/increases automatic sceen refreshing time (goes from 0.1 seconds to 60 seconds on a scale)  
P pauses/unpauses the program completely  
A turns on/off automatic screen refreshing  
G generates a new screen immediately (can be slow with a large screen)

C toggles through preset colour modes  
ARROW KEYS - resizes the window  
F toggles fullscreen mode (window resizing is bad in this mode)

S saves the current screen to an image (saved in /screenshots as a png file with a timestamp filename)  
W randomly selects another control  
ESCAPE ends the program

NOTE: most keys can be held down to be rapidly used, including saving!

History
=======
UniMazeGenEX DingoMode 2015 was originally made in 2013, when it was humbly known as Pipes Generator 2013. It had a short lived second incarnation as UniDisplayEX 2014, but was stopped short due to problems I could not figure out. Hopefully I can get a better handle on it today!

changelog.txt

v.07 Changes:
- prepared for publishing on github
- removed "totally wild" colour set
- updated program name
- added two new character sets

v0.6 Changes
- groups of characters to display can be cycled through
- relevant key for above added
- added single pipes, card suits
- updated font to courier, from consolas
- newline now adds 8 spaces instead of full lines

Marhc 12, 2015: v0.7 released!
September 25, 2013: v0.6 released!

runme.py
#!/usr/bin/python

# a program that generates and displays random series of unicode pipe characters
# with many options provided for the user to change how they see everything!

# un-comment for compiling [currently broken!]
#import pygame._view
import pygame
pygame.font.init()
from pygame.locals import *

# local pygcurse file
import pygcurse
# local settings file
from settings import *
# python libraries needed
import random, time, sys, copy, os

window = pygcurse.PygcurseWindow()
# disables autoupdate to prevent flickering
window.autoupdate = False

FPS = 30
# clock for limiting program speed
CLOCK = pygame.time.Clock()
pygame.key.set_repeat(200, 50)


def quit():
    # ends the program completely
    pygame.quit()
    sys.exit()


def copy_default_settings():
    # creates a copy of the default settings dict

    settings = {}
    for key, value in DEFAULT_SETTINGS.items():
        settings[key] = copy.deepcopy(value)
    return settings


def set_window_properties(settings):
    # changes the window size to any given values

    global window

    # sets the window size
    window.resize(settings['width'], settings['height'])
    # sets up the fonts
    window.font = pygame.font.SysFont(settings['font'], settings['fontSize'])
    # sets the caption
    pygame.display.set_caption(settings['caption'])


def set_colors(colourInfo):
    # sets the colours for the window

    fg = ()
    bg = ()

    # fg layer, bg layer
    for layer in [0, 2]:
        # R, G, and B channels
        for channel in range(len(colourInfo[layer])):

            minRoll = colourInfo[layer][channel]
            maxRoll = colourInfo[layer + 1][channel]

            roll = random.randint(minRoll, maxRoll)

            if layer == 0: fg += (roll,)
            else: bg += (roll,)

    window.setscreencolors(fg, bg)


def generate(chars, width, height):
    # generates the output of random characters from the list

    # output generated, forced to use a unicode string
    output = u''
    # adds data line by line, to avoid displaying too many chars
    newLine = u''
    numWanted = width * height

    # if there are no chars to select, blank the screen
    if len(chars) == 0:
        output += ' ' * (width * height)

    while len(output) < numWanted:

        roll = random.randint(0, len(chars) - 1)
        char = chars[roll]

        newLine += char

        if len(newLine) >= width:
            # cuts off the extra characters
            output += newLine[:width]
            # puts the extra characters onto the next line
            newLine = newLine[width:]

    return output


def user_options(keyEvent, settings, nextScreen):
    # runs through and applies options enacted by the user

    # tries to convert the key event into a string character
    char = pygcurse.interpretkeyevent(keyEvent)
    # if it is convertible, it sets it to the key
    if char: key = char.upper()
    # otherwise, it just uses the keyEvent's key
    else: key = keyEvent.key

    # if the screen should be refreshed
    refresh = False
    # DEBUG! if the list of characters to be displayed should be shown
    showDisplayChars = False

    # chooses a totally random option
    if key == WILDCARD:
        roll = random.randint(0, len(WILD_KEYS) - 1)
        key = WILD_KEYS[roll]
        print key

    # pauses/unpauses the program
    if key == PAUSE:
        settings['pause'] = not settings['pause']

    # turns automatic screen refresh on/off
    elif key == AUTOREFRESH:
        settings['autoRefresh'] = not settings['autoRefresh']


    # time controls
    elif key in TIME_KEYS:

        position = SPEED_VALUES.index(settings['speed'])

        # if the action is slowing down the refresh rate
        if key == TIME_KEYS[0] or key == K_EQUALS and (KMOD_SHIFT):
            position += 1
            action = '+'
        # or if the action is speeding it up
        else:
            position -= 1
            action = '-'

        # moves position around to valid values
        if position < 0: position = 0
        if position >= len(SPEED_VALUES): position = len(SPEED_VALUES) - 1

        settings['speed'] = SPEED_VALUES[position]
        speed = settings['speed']

        # the time to the next screen is larger than the new speed
        if action == '-' and nextScreen > speed:
            nextScreen = speed
        elif action == '+' and nextScreen < speed:
            difference = speed - nextScreen
            nextScreen += difference / 2


    # changes the current character set to the next
    elif key == CHANGE_CHAR_SET:

        # gets the numeric position of the current colour mode
        current = CHAR_SETS.index(settings['charSet'])
        # finds the position of the /next/ colour mode, going back to 1 if len
        settings['charSet'] = CHAR_SETS[ (current + 1) % len(CHAR_SETS) ]
        # puts in all of the new chars
        settings['charsToUse'] = settings['charSet'].values()

        showDisplayChars = True
        refresh = True

    # blanks chars to display
    elif key == CLEAR_CHARS:

        settings['charsToUse'] = []
        showDisplayChars = True
        refresh = True

    # selects/deselects a character for displaying
    elif key in settings['charSet'] or key in EXTRACHARS:

        char = settings['charSet'].get(key, None)
        if char == None: char = EXTRACHARS[key]

        # removes the selected 
        if char in settings['charsToUse']:
            settings['charsToUse'].remove(char)
        else:
            settings['charsToUse'].append(char)

        showDisplayChars = True
        refresh = True

    # selects/deselects only one charater for displaying
    elif key in SINGLECHARS:

        # retrives the corresponding un-shifted key, if it exists
        char = settings['charSet'].get(SINGLECHARS[key])

        # makes sure the selection is valid
        if char != None:
            settings['charsToUse'] = [char]
            showDisplayChars = True
            refresh = True

    # DEBUG TERMINAL OUTPUT
    if showDisplayChars:

        data = []
        for k, v in settings['charSet'].items():
            if v in settings['charsToUse']:
                data.append(k)
        print data


    # toggles what colour mode the screen is in
    elif key == COLOURMODE_TOGGLE:

        # gets the numeric position of the current colour mode
        current = COLOURMODENAMES.index(settings['colourMode'])
        # finds the position of the /next/ colour mode, going back to 1 if len
        settings['colourMode'] = COLOURMODENAMES[ (current + 1) % len(COLOURMODENAMES) ]

        refresh = True


    # toggles the full screen setting
    elif key == FULLSCREEN:

        settings['fullscreen'] = not settings['fullscreen']
        window.fullscreen = settings['fullscreen']

        # updates the window so it will be immediately displayed in full screen
        window.update()

    # resizes the window
    elif key in RESIZE_KEYS:

        # defaults to nothing, in case an action isn't possible
        change = (0, 0)

        # saves reference space
        action = RESIZE_KEYS.index(key)

        # creates the appropiate difference tuple, if possible
        if action == 0 and settings['width'] > 1: change = (-1, 0)
        elif action == 1 and settings['width'] < MAXWIDTH: change = (1, 0)
        elif action == 2 and settings['height'] > 1: change = (0, -1)
        elif action == 3 and settings['height'] < MAXHEIGHT: change = (0, 1)

        # applies the size changes
        settings['width'] += change[0]
        settings['height'] += change[1]

        set_window_properties(settings)
        refresh = True

    # saves the screen to an image
    elif key == SCREENSHOT:

        # makes the screenshot folder if it doesn't already exist
        if not os.path.exists(SCREENSHOT_LOCATION):
            os.makedirs(SCREENSHOT_LOCATION)

        # removes period from time.time() result
        filename = str(time.time()).replace('.', '')
        filename = "./%s/%s%s" % (SCREENSHOT_LOCATION, filename, SCREENSHOT_FILETYPE)

        pygame.image.save(window._surfaceobj, filename)


    # user wants to refresh the screen, or it needs to be done now
    if key == GENERATE or refresh:
        nextScreen = 0

    return settings, nextScreen


def display_screen(settings):
    # displays a screen of characters to the user
    
    nextScreen = settings['speed']

    set_colors(COLOURMODES[ settings['colourMode'] ])

    # blanks the window display
    window.fill('')
    # resets cursor position
    window.cursorx = 0
    window.cursory = 0

    # generates extra output, just to be sure to fill the screen
    output = generate(settings['charsToUse'], settings['width'], settings['height'])

    # writes output to the window
    window.write(output, autoScroll=False)
    # finally updates the window, displaying it to the user
    window.update()

    return nextScreen

def main():
    # the main program, runs everything from here

    # if all variables need to be set/reset
    reset = True

    # time until the next screen is displayed automatically
    nextScreen = 0
    # the system time when the screen was changed last
    lastTime = 0

    # loops until user quit
    while True:

        # if the settings need to be re-initialized
        if reset:
            reset = False
            settings = copy_default_settings()
            set_window_properties(settings)
            # force the next screen to display
            nextScreen = 0
        
        # slows down processing speed a little
        CLOCK.tick(FPS)

        # if the screen should automatically refresh
        if settings['autoRefresh']:

            # gets the difference in time since the last loop
            timeDiff = time.time() - lastTime
            # applies time difference to the time remaining
            nextScreen -= timeDiff
            # saves the exact time for the next loop
            lastTime = time.time()

        # if it is time to display another screen
        if nextScreen <= 0 and not settings['pause']:
            nextScreen = display_screen(settings)

        for event in pygame.event.get():

            # user quit
            if event.type == pygame.QUIT or (event.type == KEYDOWN and event.key == K_ESCAPE):
                quit()

            # if the user pressed a key
            if event.type == KEYDOWN:

                print pygcurse.interpretkeyevent(event)
                
                # looks through user options
                settings, nextScreen = user_options(event, settings, nextScreen)

                # reset settings to def
                if event.key == RESET_SETTINGS:
                    reset = True

# runs the program
main()

settings.py
#!/usr/bin/python

# settings for runme.py
# CURRENTLY VERSION 0.7.0!

import pygame
from pygame.locals import *


# TODO: (! means finished- keep track of controls, ~ means SORTA working)

#~ hold down keys (ie (W)ildcard)
    # ARROW KEYS can't be held down??? what's wrong with resize?

#! R to reset all values to default
#! ESC to quit

#! `1234567890 to toggle chars
#! ~!@#$%^&*() keys to select one of above's corresponding chars
#! others to toggle: TAB, SPACE, ENTER
#! B to change character set
#  ? to select all characters
#! BACKSPACE to display nothing
#  ? BACKSPACE toggles display none <> display all

#! - and + to decrease and increase the rate of automatic change, using a scaling system
#! P to pause/continue
#! A to turn on/off automatic refresh mode

#! ARROW KEYS increase / decrease windows' X/Y axises
#! F for full screen
    # AGAIN, RESIZE function is awful here!

#! C to toggle colour mode
#! G to instantly generate a new layout
#! S to save the screen to an image

#! W (Wildcard) to select a mostly random action

#  ? for help/controls (?)
#  ? for freezing colour changes

# add more colour modes
# interpret keypad ? (need to change pygcurse interpret key event)
# why is resize() so annoying

# BIG THINGS:
# M to select these new modes
# U to undo the last selection (!)
# totally wild tile mode- EVERY square is randomly changed
# continuous update mode- change each square randomly, constantly

# BIG WORLD version, where you can scroll around a giant random layout!
# implement an introductory / explanatory menu ?
# [?] key to go back to menu / resume
# seperate window for information output ?


# maximum width allowed
MAXWIDTH = 150
# maximum height allowed
MAXHEIGHT = 75


# dict of preset colour modes, ordered by range of colours
# each position of three refers to the RGB values to be used
# fg-min, fg-max, bg-min, bg-max - fg is 'text', bg is 'fill'
COLOURMODES = {
    #'template': ( (100, 100, 100), (255, 255, 255), (255, 255, 255), (100, 100, 100) )
    'default': ( (128,) * 3, (255,) * 3, (0,) * 3, (127,) * 3 ),

    'bright red': ( (50, 20, 20), (127, 30, 30), (200, 30, 30), (240, 45, 45) ),
    'bright green': ( (20, 50, 20), (30, 127, 30), (30, 200, 30), (45, 240, 45) ),
    'bright blue': ( (20, 20, 50), (30, 30, 127), (30, 30, 200), (45, 45, 240) ),

    'opal': ( (205,) * 3, (225,) * 3, (135,) * 3, (155,) * 3 ),
    'murky': ( (0,) * 3, (30,) * 3, (15,) * 3, (45,) * 3 )
}

# ordered list for colour modes to go through
COLOURMODENAMES = ['default', 'bright red', 'bright green', 'bright blue', 'opal', 'murky']

# list of speed settings to use
SPEED_VALUES = [x / 10.0 for x in range(1, 6)] + [x / 100.0 for x in range(75, 200, 25)] + \
                [x / 100.0 for x in range(200, 500, 50)] + [x for x in range(5, 10)] + \
                [x for x in range(10, 70, 10)]


# various single keys are below

# keys that can be randomly hit
GENERATE = 'G'
COLOURMODE_TOGGLE = 'C'
CHANGE_CHAR_SET = 'B'

# keys that must be manually used
PAUSE = 'P'
WILDCARD = 'W'
FULLSCREEN = 'F'
AUTOREFRESH = 'A'
CLEAR_CHARS = K_BACKSPACE
# NEEDS to be K_r because it is checked seperately from other keys
RESET_SETTINGS = K_r

# keys with a special extra thing
SCREENSHOT = 'S'
SCREENSHOT_FILETYPE = '.PNG'
SCREENSHOT_LOCATION = 'screenshots'


# keys for increasing/decreasing the width/height of the program window
RESIZE_KEYS = [K_LEFT, K_RIGHT, K_UP, K_DOWN]

# keys for controlling the rate of automatic refreshing
TIME_KEYS = ['+', '-']


# dict of keys, and their shifted states, for single character selections
SINGLECHARS = {
    '~': '`', '!': '1', '@': '2', '#': '3', '$': '4', '%': '5',
    '^': '6', '&': '7', '*': '8', '(': '9', ')': '0'
}


# dicts of chars to use, referenced by the key that toggles them

# two spaced apart pipe-like lines
DOUBLEPIPES = {
        # four-way connector
        '`': u"\u256C",
        # corner pieces: top-left, top-right, bottom-left, bottom-right
        '1': u"\u2554", '2': u"\u2557", '3': u"\u255A", '4': u"\u255D",
        # t-sections: no top, no right, no bottom, no left
        '5': u"\u2566", '6': u"\u2563", '7': u"\u2569", '8': u"\u2560",
        # two-way connectors: top-bottom, left-right; four-way connector
        '9': u"\u2550", '0': u"\u2551"
}

# single pipe-like lines
SINGLEPIPES = {
        # four-way connector
        '`': u"\u253C",
        # corner pieces: top-left, top-right, bottom-left, bottom-right
        '1': u"\u250C", '2': u"\u2510", '3': u"\u2514", '4': u"\u2518",
        # t-sections: no top, no right, no bottom, no left
        '5': u"\u252C", '6': u"\u2524", '7': u"\u2534", '8': u"\u251C",
        # two-way connectors: top-bottom, left-right; four-way connector
        '9': u"\u2502", '0': u"\u2500"
}

# playing card suits
CARDSUITS = {
        # black filled: heart, diamond, club, spades
        '`': u"\u2665", '1': u"\u2666", '2': u"\u2663", '3': u"\u2660",
        # white filled: heart, diamond, club, spades - WON'T DISPLAY :(
        #'4': u"\u2661",'5': u"\u2662", '6': u"\u2667", '7': u"\u2664",
        # nothing... yet
        #'8': u"", '9': u"", '0': u""
}

# just some numbers
NUMBERS = {
    '`': u"`", '1': u"1", '2': u"2", '3': u"3",
    '4': u"4", '5': u"5", '6': u"6", '7': u"7", 
    '8': u"8", '9': u"9", '0': u"0"
}

# shift row of numbers
SHIFT_ROW = {
    '`': u"~", '1': u"!", '2': u"@", '3': u"#",
    '4': u"$", '5': u"%", '6': u"^", '7': u"&", 
    '8': u"*", '9': u"(", '0': u")"
}

# curved corners and some diagonal lines... don't seem to work???
CURVES_DIAGONALS = {
    # curved corners: top-left, top-right, bottom-left, bottom-right
    '`': u"\u256D", '1': u"\u256E", '2': u"\u256F", '3': u"\u2570",
    # diagonals: bottom-left to upper-right, upper-left to bottom-right
    '4': u"\u2571", '5': u"\u2572",
    # four way diagonal
    '6': u"\u2573",
    # nothing... yet
    #'7': u"\u", '8': u"\u", '9': u"\u", '0': u"\u"
}

# the space, newline, and tab characters are always included
EXTRACHARS = {' ': " ", '\t': " " * 4, '\r': " " * 8}

CHAR_SETS = [DOUBLEPIPES, SINGLEPIPES, CARDSUITS, 
            NUMBERS, SHIFT_ROW]


# refers to all keys that would be interesting to be randomly pressed
WILD_KEYS = SINGLECHARS.values() + TIME_KEYS #+ RESIZE_KEYS
WILD_KEYS += [GENERATE, COLOURMODE_TOGGLE, CHANGE_CHAR_SET, ' ']


# stores the default settings
DEFAULT_SETTINGS = {

    # program window height / x size
    'width': 80,
    # program window width / y size
    'height': 25,
    # window caption
    'caption': "UNIDISPLAYEX 2015 - a Nikki production",
    # fullscreen status
    'fullscreen': False,

    # displayed font type
    'font': 'courier',
    # displayed font size
    'fontSize': 15,

    # how long, in seconds, until generating a new screen
    'speed': 4,
    # if the screen should automatically refresh
    'autoRefresh': True,
    # if the screen should never refresh
    'pause': False,

    # chosen arrangement of colours
    'colourMode': COLOURMODENAMES[0],

    # current char set to be used
    'charSet': CHAR_SETS[0],
    # list of chars to display to the screen
    'charsToUse': CHAR_SETS[0].values()
}